README.md
# baidusms

`baidusms` package provides Go support for Baidu SMS API.

This Project is non-official.

The document of Baidu SMS API is [here](https://cloud.baidu.com/doc/SMS/s/3jwvxrwjx)

The document of the generation of `Authorization` header is [here](https://cloud.baidu.com/doc/Reference/s/njwvz1yfu)

# Example

```go
package main

import (
	"errors"
	"log"

	"github.com/iftechio/baidusms/v2"
)

func main() {
	sms := baidusms.BaiduSMS{
		AccessKey: "YOUR_ACCESS_KEY",
		SecretKey: "YOUR_SECRET_KEY",
		Region:    "bj",
	}
	// example template "Your SMS code is ${code}, expires in ${hour} hours"
	contentVar := map[string]string{
		"code": "1234",
		"hour": "2",
	}
	resp, err := sms.SendSMSCode("YOUR_INVOKE_ID", "17612233344", "YOUR_TEMPLATE_CODE", contentVar)
	if err != nil {
		var sendError *baidusms.ErrSendFail
		// go 1.13
		if errors.As(err, &sendError) {
			switch sendError.APICode {
			case "4621":
				// 4621 手机号配额异常
				log.Printf("baidusms single number frequency exceed")

			case "4503":
				// 4503, 手机号码格式不正确
				log.Printf("baidusms single number format error")

			}
		}
		log.Fatalf("error %s ", err)

	}
	log.Printf("%v", resp)
}
```


CHANGELOG.md
# Changelog

## 2.0.3
1. Add timeout

## 2.0.2

1. Custom error in return value use pointer.

## 2.0.1

1. [BREAKING] Return type of SendSMSCode changes to pointer *SuccessResponse.
2. Use custom error struct instead of plain string error.


auth_test.go
package baidusms

import (
	"net/http"
	"net/url"
	"testing"
	"time"
)

// TestHash tests HMAC SHA256 hash function
func TestHash(t *testing.T) {
	data := "bce-auth-v1/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/2015-04-27T08:23:49Z/1800"
	key := "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
	result := "1d5ce5f464064cbee060330d973218821825ac6952368a482a592e6615aef479"
	testResult := hash(data, key)
	if result == testResult {
		t.Log("hash test success")
	} else {
		t.Error("hash test fail")
	}
}

// TestStringNormalize tests string normalize
// rule: https://cloud.baidu.com/doc/Reference/s/njwvz1yfu
func TestStringNormalize(t *testing.T) {
	data1 := "NFzcPqhviddjRNnSOGo4rw=="
	data2 := "text/plain"
	data3 := "Mon, 27 Apr 2015 16:23:49 +0800"

	result1 := "NFzcPqhviddjRNnSOGo4rw%3D%3D"
	result2 := "text%2Fplain"
	result3 := "Mon%2C%2027%20Apr%202015%2016%3A23%3A49%20%2B0800"

	testRes1 := stringNormalize(data1, true)
	testRes2 := stringNormalize(data2, true)
	testRes2WithoutSlash := stringNormalize(data2, false)
	testRes3 := stringNormalize(data3, true)

	errorMsg := "string normalize test fail %s"
	if testRes1 != result1 {
		t.Errorf(errorMsg, testRes1)
	}
	if testRes2 != result2 {
		t.Errorf(errorMsg, testRes2)
	}
	if testRes2WithoutSlash != data2 {
		t.Errorf(errorMsg, testRes2WithoutSlash)
	}
	if testRes3 != result3 {
		t.Errorf(errorMsg, testRes3)
	}

	t.Log("string normalize test success")
}

func TestQueryStringCanonicalization(t *testing.T) {
	qs := url.Values{
		"text":   []string{},
		"text1":  []string{"测试"},
		"text10": []string{"test"},
	}
	testRes := queryStringCanonicalization(qs)
	if testRes == "text10=test&text1=%E6%B5%8B%E8%AF%95&text=" {
		t.Log("qs Canonicalization test success")
	} else {
		t.Errorf("qs Canonicalization test fail %s", testRes)
	}
}

func TestHeadersCanonicalization(t *testing.T) {
	headers := http.Header{
		"Host":           []string{"bj.bcebos.com"},
		"Date":           []string{"Mon, 27 Apr 2015 16:23:49 +0800"},
		"Content-Type":   []string{"text/plain"},
		"Content-Length": []string{"8"},
		"Content-Md5":    []string{"NFzcPqhviddjRNnSOGo4rw=="},
		"x-bce-date":     []string{"2015-04-27T08:23:49Z"},
	}
	res, signedHeaders := headersCanonicalization(headers)
	if res == "content-length:8\ncontent-md5:NFzcPqhviddjRNnSOGo4rw%3D%3D\ncontent-type:text%2Fplain\nhost:bj.bcebos.com\nx-bce-date:2015-04-27T08%3A23%3A49Z" && signedHeaders == "content-length;content-md5;content-type;host;x-bce-date" {
		t.Log("header Canonicalization test success")
	} else {
		t.Errorf("header Canonicalization test fail %s %s", res, signedHeaders)
	}
}

func TestGenerateAuthorization(t *testing.T) {
	method := "PUT"
	headers := http.Header{
		"Host":           []string{"bj.bcebos.com"},
		"Date":           []string{"Mon, 27 Apr 2015 16:23:49 +0800"},
		"Content-Type":   []string{"text/plain"},
		"Content-Length": []string{"8"},
		"Content-Md5":    []string{"NFzcPqhviddjRNnSOGo4rw=="},
		"x-bce-date":     []string{"2015-04-27T08:23:49Z"},
	}
	// body := "Example"
	qs := url.Values{
		"partNumber": []string{"9"},
		"uploadId":   []string{"a44cc9bab11cbd156984767aad637851"},
	}
	path := "/v1/test/myfolder/readme.txt"
	auth := auth{
		AccessKey: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
		SecretKey: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
	}
	currentTime, _ := time.Parse(time.RFC3339, "2012-11-01T22:08:41Z")
	testRes := auth.generateAuthorization(method, path, headers, qs, currentTime)
	if testRes == "bce-auth-v1/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/2012-11-01T22:08:41Z/1800/content-length;content-md5;content-type;host;x-bce-date/b47ae03e099e2a1553ef56fbba0c23fd5b632c99bc118df7f652b0fe1b4c5988" {
		t.Log("auth test success")
	} else {
		t.Errorf("auth test fail %s", testRes)
	}
}


auth.go
package baidusms

import (
	"crypto/hmac"
	"crypto/sha256"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

// Auth is auth module of baidu api authenticator
type auth struct {
	AccessKey string
	SecretKey string
}

// getCanonicalTime returns RFC3339 style time string
// current param provides convenience for testing
func getCanonicalTime(current time.Time) (isoStr string) {
	utc, _ := time.LoadLocation("UTC")
	isoStr = current.In(utc).Format(time.RFC3339)
	return
}

func hash(data string, key string) string {
	h := hmac.New(sha256.New, []byte(key))
	io.WriteString(h, data)
	return fmt.Sprintf("%x", h.Sum(nil))
}

func stringNormalize(str string, encodingSlash bool) string {
	var b strings.Builder
	for _, ch := range str {
		if (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || ch == '_' || ch == '-' || ch == '~' || ch == '.' {
			b.WriteRune(rune(ch))
		} else if ch == '/' {
			if encodingSlash {
				b.WriteString("%2F")
			} else {
				b.WriteRune(rune(ch))
			}
		} else {
			// query escape turn space to +, but baidu api turn space to %20
			// issue: https://github.com/golang/go/issues/4013
			b.WriteString(strings.ReplaceAll(url.QueryEscape(string(ch)), "+", "%20"))
		}
	}
	return b.String()
}

func uriCanonicalization(path string) string {
	return path
}

func queryStringCanonicalization(query url.Values) string {
	queryPatternArray := []string{}
	for k, values := range query {
		if strings.ToLower(k) == "authorization" {
			continue
		}
		if len(values) > 0 {
			for _, v := range values {
				queryPatternArray = append(queryPatternArray, fmt.Sprintf("%s=%s", k, stringNormalize(v, true)))
			}
		} else {
			queryPatternArray = append(queryPatternArray, fmt.Sprintf("%s=", k))
		}
	}
	sort.Strings(queryPatternArray)
	return strings.Join(queryPatternArray, "&")
}

func headersCanonicalization(headers http.Header) (string, string) {
	headersToSign := map[string]bool{"host": true, "content-md5": true, "content-length": true, "content-type": true}
	canonicalHeaders := []string{}
	signedHeaders := []string{}
	for k, values := range headers {
		k = strings.ToLower(k)
		if headersToSign[k] || strings.HasPrefix(k, "x-bce-") {
			signedHeaders = append(signedHeaders, k)
			for _, v := range values {
				if v != "" {
					canonicalHeaders = append(canonicalHeaders, fmt.Sprintf("%s:%s", stringNormalize(k, true), stringNormalize(v, true)))
				}
			}
		}
	}
	sort.Strings(canonicalHeaders)
	sort.Strings(signedHeaders)
	return strings.Join(canonicalHeaders, "\n"), strings.Join(signedHeaders, ";")
}

var defaultExpirationInSeconds = 1800

//  timestamp int64, expirationInSeconds int, headersToSign []string 这个三个不支持了
func (auth auth) generateAuthorization(httpMethod string, path string, headers http.Header, query url.Values, currentTime time.Time) string {
	rawSessionKey := fmt.Sprintf("bce-auth-v1/%s/%s/%d", auth.AccessKey, getCanonicalTime(currentTime), defaultExpirationInSeconds)
	sessionKey := hash(rawSessionKey, auth.SecretKey)
	canonicalURI := uriCanonicalization(path)
	canonicalQueryString := queryStringCanonicalization(query)
	canonicalHeaders, signedHeaders := headersCanonicalization(headers)
	rawSignature := fmt.Sprintf("%s\n%s\n%s\n%s", httpMethod, canonicalURI, canonicalQueryString, canonicalHeaders)
	signature := hash(rawSignature, sessionKey)
	if len(signedHeaders) > 0 {
		return fmt.Sprintf("%s/%s/%s", rawSessionKey, signedHeaders, signature)
	}
	return fmt.Sprintf("%s//%s", rawSessionKey, signature)
}


baidusms.go
package baidusms

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// BaiduSMS is config of sms service. AccessKey, SecretKey, Region should be provided
// Region is one of "bj" and "gz"
type BaiduSMS struct {
	AccessKey string
	SecretKey string
	Region    string
}

// SuccessResponse is success body of baidu response
type SuccessResponse struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	RequestID string `json:"requestId"`
}

// ErrSendFail is fail body of baidu response
type ErrSendFail struct {
	HTTPCode  int
	APICode   string
	Message   string
	RequestID string
}

func (e *ErrSendFail) Error() string {
	return fmt.Sprintf("Baidu SMS API error, httpcode: %d, code: %s, message: %s, requestID: %s",
		e.HTTPCode, e.APICode, e.Message, e.RequestID)
}

var (
	// Version of baidusms
	Version = "2.0.3"
)

func (bd BaiduSMS) sendRequest(method string, path string, body string) (*SuccessResponse, error) {
	now := time.Now()
	auth := auth{bd.AccessKey, bd.SecretKey}
	var host string
	if strings.ToLower(bd.Region) == "gz" {
		host = "sms.gz.baidubce.com"
	} else {
		host = "sms.bj.baidubce.com"
	}
	targetURL := fmt.Sprintf("https://%s%s", host, path)
	req, err := http.NewRequest(method, targetURL, strings.NewReader(body))
	req.Header.Add("User-Agent", fmt.Sprintf("bce-sdk-go/%s", Version))
	req.Header.Add("Host", host)
	req.Header.Add("Connection", "close")
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Content-Length", strconv.FormatInt(req.ContentLength, 10))
	req.Header.Add("x-bce-date", getCanonicalTime(now))
	sum := sha256.Sum256([]byte(body))
	req.Header.Add("x-bce-content-sha256", hex.EncodeToString(sum[:]))
	headers := req.Header
	req.Header.Add("Authorization", auth.generateAuthorization(method, path, headers, url.Values{}, now))
	client := &http.Client{
		Timeout: 10 * time.Second,
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		bodyBytes, readErr := ioutil.ReadAll(resp.Body)
		if readErr != nil {
			return nil, readErr
		}
		var s SuccessResponse
		err = json.Unmarshal(bodyBytes, &s)
		if err != nil {
			return nil, err
		}
		if s.Code != "1000" {
			// only 1000 is correct
			return nil, &ErrSendFail{
				HTTPCode:  resp.StatusCode,
				APICode:   s.Code,
				Message:   s.Message,
				RequestID: s.RequestID,
			}
		}
		return &s, nil
	}
	return nil, &ErrSendFail{
		HTTPCode: resp.StatusCode,
	}
}

type requestBody struct {
	InvokeID     string            `json:"invokeId"`
	PhoneNumber  string            `json:"phoneNumber"`
	TemplateCode string            `json:"templateCode"`
	ContentVar   map[string]string `json:"contentVar"`
}

// SendSMSCode will call HTTP request to Baidu API to send a sms
func (bd BaiduSMS) SendSMSCode(invokeID string, mobilePhoneNumber string, templateCode string, contentVar map[string]string) (*SuccessResponse, error) {
	path := "/bce/v2/message"
	body := requestBody{invokeID, mobilePhoneNumber, templateCode, contentVar}
	bodyStr, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	return bd.sendRequest("POST", path, string(bodyStr))
}


example/main.go
package main

import (
	"errors"
	"log"

	"github.com/iftechio/baidusms/v2"
)

func main() {
	sms := baidusms.BaiduSMS{
		AccessKey: "YOUR_ACCESS_KEY",
		SecretKey: "YOUR_SECRET_KEY",
		Region:    "bj",
	}
	// content var is related your sms template
	// we use example template "Your SMS code is ${code}, expires in ${hour} hours"
	// so contentVar should contain code and hour
	contentVar := map[string]string{
		"code": "1234",
		"hour": "2",
	}
	resp, err := sms.SendSMSCode("YOUR_INVOKE_ID", "17612233344", "YOUR_TEMPLATE_CODE", contentVar)
	if err != nil {
		var sendError *baidusms.ErrSendFail
		// go 1.13
		if errors.As(err, &sendError) {
			switch sendError.APICode {
			case "4621":
				// 4621 手机号配额异常
				log.Printf("baidusms single number frequency exceed")

			case "4503":
				// 4503, 手机号码格式不正确
				log.Printf("baidusms single number format error")

			}
		}
		log.Fatalf("error %s ", err)

	}
	log.Printf("%v", resp)
}
